File: app/src/main/java/com/example/pam_app/model/Bucket.java
package com.example.pam_app.model;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.Date;
import java.util.EnumSet;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

public class Bucket implements Serializable {
    public final String title;
    public final Date dueDate;
    public final BucketType bucketType;
    public final double target;
    public final List<BucketEntry> entries;
    public final List<BucketEntry> oldEntries;
    public final Integer id;
    public final String imagePath;
    public final boolean isRecurrent;

    public Bucket(String title, Date dueDate, BucketType bucketType, double target,
                  List<BucketEntry> entries, boolean isRecurrent) {
        this.title = title;
        this.dueDate = dueDate;
        this.bucketType = bucketType;
        this.target = target;
        this.entries = new LinkedList<>();
        this.oldEntries = new LinkedList<>();
        this.id = null;
        this.imagePath = null;
        this.isRecurrent = isRecurrent;
        initEntries(entries);
    }

    public Bucket(String title, Date dueDate, BucketType bucketType, double target,
                  List<BucketEntry> entries, Integer id, String imagePath, boolean isRecurrent) {
        this.title = title;
        this.dueDate = dueDate;
        this.bucketType = bucketType;
        this.target = target;
        this.entries = new LinkedList<>();
        this.oldEntries = new LinkedList<>();
        this.id = id;
        this.imagePath = imagePath;
        this.isRecurrent = isRecurrent;
        initEntries(entries);
    }

    public Bucket(String title, Date dueDate, BucketType bucketType, double target, boolean isRecurrent) {
        this.title = title;
        this.dueDate = dueDate;
        this.bucketType = bucketType;
        this.target = target;
        this.entries = new LinkedList<>();
        this.oldEntries = new LinkedList<>();
        this.id = null;
        this.imagePath = null;
        this.isRecurrent = isRecurrent;
    }

    public Bucket(String title, Date dueDate, BucketType bucketType, double target, String imagePath, boolean isRecurrent) {
        this.title = title;
        this.dueDate = dueDate;
        this.bucketType = bucketType;
        this.target = target;
        this.entries = new LinkedList<>();
        this.oldEntries = new LinkedList<>();
        this.id = null;
        this.imagePath = imagePath;
        this.isRecurrent = isRecurrent;
    }

    public int getProgress() {
        double total = this.getTotal();

        if (total == 0) {
            return 0;
        }
        return (int) ((total / target) * 100);
    }

    public double getTotal() {
        double total = 0;
        for (BucketEntry entry: this.entries) {
            total += entry.amount;
        }
        return total;
    }

    public String getTotalString() {
        return "" + this.getTotal() + " / " + this.target;
    }

    public String getRemainingTime() {
        Date timeNow = new Date();
        if (dueDate.getTime() > timeNow.getTime()) {
            Map<TimeUnit,Long> diff = this.computeDiff(new Date(), dueDate);
            return "" + diff.get(TimeUnit.DAYS) + "D " + diff.get(TimeUnit.HOURS) + "Hs";
        }
        return "0D 0Hs";
    }

    public Bucket setDueDate(final Date newDueDate) {
        return new Bucket(title, newDueDate, bucketType, target,
                entries, id, imagePath, isRecurrent);
    }

    private Map<TimeUnit,Long> computeDiff(Date date1, Date date2) {
        long diffInMillis = date2.getTime() - date1.getTime();
        List<TimeUnit> units = new ArrayList<>(EnumSet.allOf(TimeUnit.class));
        Collections.reverse(units);
        Map<TimeUnit,Long> result = new LinkedHashMap<>();
        long milliesRest = diffInMillis;

        for (TimeUnit unit : units) {
            long diff = unit.convert(milliesRest,TimeUnit.MILLISECONDS);
            long diffInMilliesForUnit = unit.toMillis(diff);
            milliesRest = milliesRest - diffInMilliesForUnit;
            result.put(unit,diff);
        }

        return result;
    }

    private void initEntries(List<BucketEntry> entries) {
        if (entries != null) {
            Calendar currentDate = Calendar.getInstance();
            currentDate.set(Calendar.HOUR_OF_DAY, 0);
            currentDate.set(Calendar.MINUTE, 0);
            currentDate.set(Calendar.SECOND, 0);
            currentDate.set(Calendar.DAY_OF_MONTH, 1);
            Date firstDayOfMonth = currentDate.getTime();
            for (BucketEntry e : entries) {
                if (this.isRecurrent && e.date.before(firstDayOfMonth)) {
                    this.oldEntries.add(e);
                } else {
                    this.entries.add(e);
                }
            }
        }
    }
}

File: app/src/main/java/com/example/pam_app/model/Bucket.kt
package com.example.pam_app.model

import java.io.Serializable
import java.util.*
import java.util.concurrent.TimeUnit
import kotlin.jvm.Synchronized

class Bucket : Serializable {
    @kotlin.jvm.JvmField
    val title: String?
    @kotlin.jvm.JvmField
    val dueDate: Date?
    @kotlin.jvm.JvmField
    val bucketType: BucketType
    val target: Double
    @kotlin.jvm.JvmField
    val entries: MutableList<BucketEntry?>
    @kotlin.jvm.JvmField
    val oldEntries: MutableList<BucketEntry?>
    val id: Int?
    val imagePath: String?
    @kotlin.jvm.JvmField
    val isRecurrent: Boolean

    constructor(title: String?, dueDate: Date?, bucketType: BucketType, target: Double,
                entries: List<BucketEntry?>?, isRecurrent: Boolean) {
        this.title = title
        this.dueDate = dueDate
        this.bucketType = bucketType
        this.target = target
        this.entries = LinkedList()
        oldEntries = LinkedList()
        id = null
        imagePath = null
        this.isRecurrent = isRecurrent
        initEntries(entries)
    }

    constructor(title: String?, dueDate: Date?, bucketType: BucketType, target: Double,
                entries: List<BucketEntry?>?, id: Int?, imagePath: String?, isRecurrent: Boolean) {
        this.title = title
        this.dueDate = dueDate
        this.bucketType = bucketType
        this.target = target
        this.entries = LinkedList()
        oldEntries = LinkedList()
        this.id = id
        this.imagePath = imagePath
        this.isRecurrent = isRecurrent
        initEntries(entries)
    }

    constructor(title: String?, dueDate: Date?, bucketType: BucketType, target: Double, isRecurrent: Boolean) {
        this.title = title
        this.dueDate = dueDate
        this.bucketType = bucketType
        this.target = target
        entries = LinkedList()
        oldEntries = LinkedList()
        id = null
        imagePath = null
        this.isRecurrent = isRecurrent
    }

    constructor(title: String?, dueDate: Date?, bucketType: BucketType, target: Double, imagePath: String?, isRecurrent: Boolean) {
        this.title = title
        this.dueDate = dueDate
        this.bucketType = bucketType
        this.target = target
        entries = LinkedList()
        oldEntries = LinkedList()
        id = null
        this.imagePath = imagePath
        this.isRecurrent = isRecurrent
    }

    val progress: Int
        get() {
            val total = total
            return if (total == 0.0) {
                0
            } else (total / target * 100).toInt()
        }

    val total: Double
        get() {
            var total = 0.0
            for (entry in entries) {
                total += entry!!.amount
            }
            return total
        }

    val totalString: String
        get() = "" + total + " / " + target

    val remainingTime: String
        get() {
            val timeNow = Date()
            if (dueDate!!.time > timeNow.time) {
                val diff = computeDiff(Date(), dueDate)
                return "" + diff[TimeUnit.DAYS] + "D " + diff[TimeUnit.HOURS] + "Hs"
            }
            return "0D 0Hs"
        }

    fun setDueDate(newDueDate: Date?): Bucket {
        return Bucket(title, newDueDate, bucketType, target,
                entries, id, imagePath, isRecurrent)
    }

    private fun computeDiff(date1: Date, date2: Date?): Map<TimeUnit?, Long> {
        val diffInMillis = date2!!.time - date1.time
        val units: List<TimeUnit?> = ArrayList(EnumSet.allOf(TimeUnit::class.java))
        Collections.reverse(units)
        val result: MutableMap<TimeUnit?, Long> = LinkedHashMap()
        var milliesRest = diffInMillis
        for (unit in units) {
            val diff = unit!!.convert(milliesRest, TimeUnit.MILLISECONDS)
            val diffInMilliesForUnit = unit.toMillis(diff)
            milliesRest = milliesRest - diffInMilliesForUnit
            result[unit] = diff
        }
        return result
    }

    private fun initEntries(entries: List<BucketEntry?>?) {
        if (entries != null) {
            val currentDate = Calendar.getInstance()
            currentDate[Calendar.HOUR_OF_DAY] = 0
            currentDate[Calendar.MINUTE] = 0
            currentDate[Calendar.SECOND] = 0
            currentDate[Calendar.DAY_OF_MONTH] = 1
            val firstDayOfMonth = currentDate.time
            for (e in entries) {
                if (isRecurrent && e!!.date!!.before(firstDayOfMonth)) {
                    oldEntries.add(e)
                } else {
                    this.entries.add(e)
                }
            }
        }
    }
}